File: python/features/steps/rekey_steps.py
"""Behave steps for project rekey scenarios."""

from __future__ import annotations

import subprocess
from pathlib import Path
from behave import given, then, when
import yaml

from features.steps.shared import run_cli


@given("the project is committed to git")
def given_project_committed(context: object) -> None:
    """Commit the project to git."""
    root = Path(context.working_directory)
    try:
        # Configure git for tests
        subprocess.run(
            ["git", "config", "user.email", "test@example.com"],
            cwd=root,
            capture_output=True,
            check=False,
        )
        subprocess.run(
            ["git", "config", "user.name", "Test User"],
            cwd=root,
            capture_output=True,
            check=False,
        )
        # Add and commit
        subprocess.run(["git", "add", "-A"], cwd=root, capture_output=True, check=False)
        subprocess.run(
            ["git", "commit", "-m", "test setup"],
            cwd=root,
            capture_output=True,
            check=False,
        )
    except Exception:
        pass


@given("the project directory has uncommitted changes")
def given_uncommitted_changes(context: object) -> None:
    """Create uncommitted changes in project/ directory."""
    from features.steps.shared import load_project_directory

    project_dir = load_project_directory(context)
    test_file = project_dir / "test-change.txt"
    test_file.write_text("uncommitted")


@when('I run "kanbus rekey {args}"')
def when_run_rekey(context: object, args: str) -> None:
    """Run the rekey command."""
    run_cli(context, f"kanbus rekey {args}")


@given('issue "{identifier}" is blocked by "{blocker}"')
def given_issue_blocked_by(context: object, identifier: str, blocker: str) -> None:
    """Add a blocked-by dependency to an issue."""
    from features.steps.shared import load_project_directory, read_issue_file, run_cli

    project_dir = load_project_directory(context)
    # Use kanbus dep command to add dependency
    run_cli(context, f'kanbus dep "{identifier}" blocked-by "{blocker}"')
    # Check if the command succeeded
    if context.result.exit_code != 0:
        raise AssertionError(f"Failed to add dependency: {context.result.stderr}")
    # Verify the dependency was created
    issue = read_issue_file(project_dir, identifier)
    found = False
    for dep in issue.dependencies:
        if dep.target == blocker and dep.dependency_type == "blocked-by":
            found = True
            break
    if not found:
        raise AssertionError(
            f"Dependency not created. Issue has: {[(d.target, d.dependency_type) for d in issue.dependencies]}"
        )


@given('issue "{identifier}" has a comment "{text}"')
def given_issue_has_comment(context: object, identifier: str, text: str) -> None:
    """Add a comment to an issue."""
    from features.steps.shared import run_cli

    # Use kanbus comment command to add comment
    run_cli(context, f'kanbus comment "{identifier}" "{text}"')


@then('issue "{identifier}" should be blocked by "{blocker}"')
def then_issue_blocked_by(context: object, identifier: str, blocker: str) -> None:
    """Verify an issue is blocked by another."""
    from features.steps.shared import load_project_directory, read_issue_file

    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    for dep in issue.dependencies:
        if dep.dependency_type == "blocked-by" and dep.target == blocker:
            return
    # Debug: print all dependencies
    deps_debug = [(d.target, d.dependency_type) for d in issue.dependencies]
    raise AssertionError(
        f"Issue {identifier} should be blocked by {blocker}. Found dependencies: {deps_debug}"
    )


@then('issue "{identifier}" should have a comment "{text}"')
def then_issue_has_comment(context: object, identifier: str, text: str) -> None:
    """Verify an issue has a comment with specific text."""
    from features.steps.shared import load_project_directory, read_issue_file

    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    for comment in issue.comments:
        comment_text = (
            comment.get("text") if isinstance(comment, dict) else (comment.text or "")
        )
        if text in comment_text:
            return
    raise AssertionError(f"Comment '{text}' not found in issue {identifier}")


@then('.kanbus.yml should have project_key "{key}"')
def then_project_key_in_file(context: object, key: str) -> None:
    """Verify project_key in .kanbus.yml file."""
    root = Path(context.working_directory)
    config_path = root / ".kanbus.yml"
    config_content = yaml.safe_load(config_path.read_text(encoding="utf-8"))
    assert (
        config_content.get("project_key") == key
    ), f"Expected project_key '{key}' but got '{config_content.get('project_key')}'"


@given('.kanbus.yml ends with the comment "{comment}"')
def given_config_comment(context: object, comment: str) -> None:
    """Append a comment line to .kanbus.yml."""
    config_path = Path(context.working_directory) / ".kanbus.yml"
    contents = config_path.read_text(encoding="utf-8")
    if not contents.endswith("\n"):
        contents += "\n"
    config_path.write_text(contents + comment + "\n", encoding="utf-8")


@then('.kanbus.yml should contain "{text}"')
def then_config_contains(context: object, text: str) -> None:
    """Verify .kanbus.yml contains text."""
    contents = (Path(context.working_directory) / ".kanbus.yml").read_text(
        encoding="utf-8"
    )
    assert text in contents, f"{text!r} not in .kanbus.yml:\n{contents}"

File: rust/features/steps/rekey_steps.rs
use std::fs;
use std::path::PathBuf;
use std::process::Command;

use cucumber::{given, then};
use serde_json::{json, Value};

use kanbus::file_io::load_project_directory;

use crate::step_definitions::initialization_steps::KanbusWorld;
use crate::step_definitions::issue_creation_steps::given_kanbus_project;

fn project_dir(world: &KanbusWorld) -> PathBuf {
    let cwd = world.working_directory.as_ref().expect("working directory");
    load_project_directory(cwd).expect("project dir")
}

fn issue_path(world: &KanbusWorld, issue_id: &str) -> PathBuf {
    project_dir(world)
        .join("issues")
        .join(format!("{issue_id}.json"))
}

fn read_issue(world: &KanbusWorld, issue_id: &str) -> Value {
    let contents = fs::read_to_string(issue_path(world, issue_id)).expect("read issue");
    serde_json::from_str(&contents).expect("parse issue")
}

fn write_issue(world: &KanbusWorld, issue_id: &str, issue: &Value) {
    let contents = serde_json::to_string_pretty(issue).expect("serialize issue");
    fs::write(issue_path(world, issue_id), contents).expect("write issue");
}

fn project_key(world: &KanbusWorld) -> String {
    let cwd = world.working_directory.as_ref().expect("working directory");
    let contents = fs::read_to_string(cwd.join(".kanbus.yml")).expect("read config");
    let config: serde_yaml::Value = serde_yaml::from_str(&contents).expect("parse config");
    config["project_key"].as_str().unwrap_or("").to_string()
}

fn git(world: &KanbusWorld, args: &[&str]) {
    let cwd = world.working_directory.as_ref().expect("working directory");
    let output = Command::new("git")
        .args(args)
        .current_dir(cwd)
        .output()
        .expect("run git");
    assert!(
        output.status.success(),
        "git {:?} failed: {}",
        args,
        String::from_utf8_lossy(&output.stderr)
    );
}

#[given(expr = "a Kanbus project with key {string}")]
fn given_project_with_key(world: &mut KanbusWorld, key: String) {
    given_kanbus_project(world);
    let cwd = world.working_directory.as_ref().expect("working directory");
    let config_path = cwd.join(".kanbus.yml");
    let contents = fs::read_to_string(&config_path).expect("read config");
    let mut config: serde_yaml::Value = serde_yaml::from_str(&contents).expect("parse config");
    config["project_key"] = serde_yaml::Value::String(key);
    fs::write(
        config_path,
        serde_yaml::to_string(&config).expect("serialize config"),
    )
    .expect("write config");
}

#[given("the project is committed to git")]
fn given_project_committed(world: &mut KanbusWorld) {
    git(world, &["add", "-A"]);
    git(
        world,
        &[
            "-c",
            "user.email=test@example.com",
            "-c",
            "user.name=Test User",
            "commit",
            "-q",
            "-m",
            "test setup",
        ],
    );
}

#[given("the project directory has uncommitted changes")]
fn given_uncommitted_changes(world: &mut KanbusWorld) {
    fs::write(project_dir(world).join("test-change.txt"), "uncommitted").expect("write test file");
}

#[given(expr = "issue {string} is blocked by {string}")]
fn given_issue_blocked(world: &mut KanbusWorld, issue_id: String, blocker: String) {
    let mut issue = read_issue(world, &issue_id);
    issue["dependencies"] = json!([{ "target": blocker, "type": "blocked-by" }]);
    write_issue(world, &issue_id, &issue);
}

#[given(expr = "issue {string} has a comment {string}")]
fn given_issue_has_comment(world: &mut KanbusWorld, issue_id: String, text: String) {
    let mut issue = read_issue(world, &issue_id);
    issue["comments"] = json!([{
        "id": "comment-1",
        "author": "test",
        "text": text,
        "created_at": "2026-02-11T00:00:00Z",
    }]);
    write_issue(world, &issue_id, &issue);
}

#[then(expr = "issue {string} should be blocked by {string}")]
fn then_issue_blocked(world: &mut KanbusWorld, issue_id: String, blocker: String) {
    let issue = read_issue(world, &issue_id);
    let found = issue["dependencies"].as_array().is_some_and(|deps| {
        deps.iter()
            .any(|dep| dep["target"] == blocker.as_str() && dep["type"] == "blocked-by")
    });
    assert!(found, "{issue_id} is not blocked by {blocker}: {issue}");
}

#[then(expr = "issue {string} should have description {string}")]
fn then_issue_has_description(world: &mut KanbusWorld, issue_id: String, description: String) {
    let issue = read_issue(world, &issue_id);
    assert_eq!(issue["description"].as_str().unwrap_or(""), description);
}

#[then(expr = "issue {string} should have a comment {string}")]
fn then_issue_has_comment(world: &mut KanbusWorld, issue_id: String, text: String) {
    let issue = read_issue(world, &issue_id);
    let found = issue["comments"].as_array().is_some_and(|comments| {
        comments
            .iter()
            .any(|comment| comment["text"].as_str() == Some(text.as_str()))
    });
    assert!(found, "{issue_id} has no comment {text:?}: {issue}");
}

#[then(expr = ".kanbus.yml should have project_key {string}")]
fn then_project_key_is(world: &mut KanbusWorld, key: String) {
    assert_eq!(project_key(world), key);
}

#[given(expr = ".kanbus.yml ends with the comment {string}")]
fn given_config_comment(world: &mut KanbusWorld, comment: String) {
    let cwd = world.working_directory.as_ref().expect("working directory");
    let config_path = cwd.join(".kanbus.yml");
    let mut contents = fs::read_to_string(&config_path).expect("read config");
    if !contents.ends_with('\n') {
        contents.push('\n');
    }
    contents.push_str(&comment);
    contents.push('\n');
    fs::write(config_path, contents).expect("write config");
}

#[then(expr = ".kanbus.yml should contain {string}")]
fn then_config_contains(world: &mut KanbusWorld, text: String) {
    let cwd = world.working_directory.as_ref().expect("working directory");
    let contents = fs::read_to_string(cwd.join(".kanbus.yml")).expect("read config");
    assert!(
        contents.contains(&text),
        "{text:?} not in .kanbus.yml:\n{contents}"
    );
}
